test/cbmc/stubs/core_sntp_stubs.c

```diff
@@ -0,0 +1,183 @@
+        .fractions = 1000      \
+    }
+
+/* An exclusive bound on the times that the NetworkInterfaceSendStub will be
+ * invoked before returning a loop terminating value. This is usually defined
+ * in the Makefile of the harnessed function. */
+#ifndef MAX_NETWORK_SEND_TRIES
+    #define MAX_NETWORK_SEND_TRIES    3
+#endif
+
+/* An exclusive bound on the times that the NetworkInterfaceReceiveStub will
+ * return an unbound value. At this value and beyond, the
+ * NetworkInterfaceReceiveStub will return zero on every call. */
+#ifndef MAX_NETWORK_RECV_TRIES
+    #define MAX_NETWORK_RECV_TRIES    4
+#endif
+
+static SntpTimestamp_t testTime = TEST_TIMESTAMP;
+
+int32_t NetworkInterfaceReceiveStub( NetworkContext_t * pNetworkContext,
+                                     uint32_t serverAddr,
+                                     uint16_t serverPort,
+                                     void * pBuffer,
+                                     size_t bytesToRecv )
+{
+    __CPROVER_assert( pBuffer != NULL,
+                      "NetworkInterfaceReceiveStub pBuffer is NULL." );
+
+    __CPROVER_assert( __CPROVER_w_ok( pBuffer, bytesToRecv ),
+                      "NetworkInterfaceReceiveStub pBuffer is writable up to bytesToRecv." );
+
+    __CPROVER_havoc_object( pBuffer );
+
+    int32_t bytesOrError;
+    static size_t tries = 0;
+
+    /* It is a bug for the application defined transport send function to return
+     * more than bytesToRecv. */
+    __CPROVER_assume( bytesOrError <= ( int32_t ) bytesToRecv );
+
+    if( tries < ( MAX_NETWORK_RECV_TRIES - 1 ) )
+    {
+        tries++;
+    }
+    else
+    {
+        bytesOrError = 0;
+    }
+
+    return bytesOrError;
+}
+int32_t NetworkInterfaceSendStub( NetworkContext_t * pNetworkContext,
+                                  uint32_t serverAddr,
+                                  uint16_t serverPort,
+                                  const void * pBuffer,
+                                  size_t bytesToSend )
+{
+    __CPROVER_assert( pBuffer != NULL,
+                      "NetworkInterfaceSendStub pBuffer is NULL." );
+
+    /* The number of tries to send the message before this invocation. */
+    static size_t tries = 1;
+
+    int32_t bytesOrError;
+
+    /* It is a bug for the application defined transport send function to return
+     * more than bytesToSend. */
+    __CPROVER_assume( bytesOrError <= ( int32_t ) bytesToSend );
+
+    /* If the maximum tries are reached, then return a timeout. In the SNTP library
+     * this stub is wrapped in a loop that will does not end until the bytesOrError
+     * returned is negative. This means we could loop possibly INT32_MAX
+     * iterations. Looping for INT32_MAX times adds no value to the proof.
+     * What matters is that the SNTP library can handle all the possible values
+     * that could be returned. */
+    if( tries < ( MAX_NETWORK_SEND_TRIES - 1 ) )
+    {
+        tries++;
+    }
+    else
+    {
+        tries = 1;
```

Comment:
Shouldn't this be set to `0` so that the next call to `Sntp_SendTimeRequest` from the proof causes this function to be called `MAX_NETWORK_SEND_TRIES` number of times?

Reply:
I have initialized this static variable as 1  on Line 97 so here also it should be set to its initial value. We can put it as 0 also but the we need to initialize it as 0. Both ways it works fine

Reply:
Having the starting value as `1` gives the incorrect semantics that a try has already been made.  It would be better that it is initialized to zero to truly represent the max number of tries allowed.